core/pom.xml

```diff
@@ -96,6 +96,14 @@ limitations under the License.
       <groupId>com.yahoo.datasketches</groupId>
       <artifactId>sketches-core</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.ow2.asm</groupId>
+      <artifactId>asm</artifactId>
```

Comment:
Is `asm` dependency really required?

Reply:
Jayway jsonpath is dependent on asm 5.0.x where elasticsearch-lang-painless uses 5.1, these two version has conflict producing NoSuchMethodError when running unit tests of calcite-elasticsearch.

What I have found is that jsonpath-2.4.0 is also working well with 5.1 version of asm, so I define this asm dependency explicitly to avoid the unit test failure.

A problem is that elasticsearch-lang-painless's asm is called `asm-debug-all` 5.1, and the jsonpath's asm is `asm` 5.1.

I am not sure the negative effect to have both of them included in dependency sets, but unit tests work well with that.

Reply:
That looks very odd.

1) `asm` is not used in Calcite code, so it should not be in `<dependency>`
2) Here's dependency tree:
```
[INFO] +- com.jayway.jsonpath:json-path:jar:2.4.0:compile
[INFO] |  \- net.minidev:json-smart:jar:2.3:compile
[INFO] |     \- net.minidev:accessors-smart:jar:1.2:compile
[INFO] |        \- org.ow2.asm:asm:jar:5.0.4:compile
```

Looking into https://github.com/netplex/json-smart-v2/blob/master/accessors-smart/src/main/java/net/minidev/asm/BeansAccess.java#L54 , I am sure it could result in memory leaks, since it keeps strong references to the Class objects.

Is `accessors-smart` truly important? Are there other alternatives?

Reply:
Thanks for your post, and you are right, the `asm-debug-all` is just a test dependency in current version of calcite, so `asm` is newly introduced compile dependency in my code. Does the `asm` itself have any problems to use?

And thanks for the notice about memory leak possibility of `accessors-smart`, I will try to deal with this problem.

Reply:
I have take a look at the code of json-smart.

The ConcurrentHashMap java type cache seems to be safe in most senarios, unless:

1. the referenced type passed to json-smart is auto generated;
2. the referenced type passed to json-smart is varing dramatically.

As we don't need to pass any type reference to json-smart, and json-path does not generate class in runtime, I think the memory leak will not be triggered in general cases. What do you think?

After this I found that Json-Path has JSON provider configurations, as Jackson is already included by calcite, I have excluded json-smart from dependency set and use Jackson JSON provider instead.

Then there will be only 1 new jar `json-path:2.4.0` added to calcite, I think this is anyway better than introducing 4 new jars.

Reply:
>Does the asm itself have any problems to use?

ASM 5 does not support Java 9/10/11.
Adding ASM 5 now is a nonsense.

> The ConcurrentHashMap java type cache seems to be safe in most senarios

`Class` instance holds a string reference to its `ClassLoader`, so just a single strong reference to `java.lang.Class` is enough to keep the whole `ClassLoader` alive.

That is why `Cache<Class>` and/or `Cache<ClassLoader>` (which is the same thing) is almost always results in a memory leak unless the code is very robust.

If you can refrain from adding `asm` as a transitive dependency, that would be way easier to reason about (== no need to invent counter-cases, etc).